File: ios/Runner/BluetoothAdapterDelegate.swift
import Foundation
import CoreBluetooth

import Flutter

/// This class represents the delegate that manages a ``CBCentralManager``.
class BluetoothAdapterDelegate : NSObject, CBCentralManagerDelegate {
    private let BLUETOOTH_SCAN_RESULT_ERROR_CODE = "BLUETOOTH_SCAN_RESULT_ERROR"
    private let BLUETOOTH_SCAN_RESULT_ERROR_MESSAGE = "A scan result returned an error."
    
    private let BLUETOOTH_UNAUTHORIZED_ERROR_CODE = "BLUETOOTH_UNAUTHORIZED"
    private let BLUETOOTH_UNAUTHORIZED_ERROR_MESSAGE = "Access to Bluetooth was denied."

    private let BLUETOOTH_UNAVAILABLE_ERROR_CODE = "BLUETOOTH_UNAVAILABLE"
    private let BLUETOOTH_UNAVAILABLE_ERROR_MESSAGE = "Bluetooth is not supported on this device."
    
    private let BLUETOOTH_PERMISSION_REQUEST_ONGOING_ERROR_CODE = "BLUETOOTH_PERMISSION_REQUEST_ONGOING"
    private let BLUETOOTH_PERMISSION_REQUEST_ONGOING_ERROR_MESSAGE = "A Bluetooth permission request is in progress."
    
    /// The internal Bluetooth manager.
    private var _bluetoothManager: CBCentralManager?
    
    /// The pending Bluetooth permission result.
    ///
    /// This result is cached when the Bluetooth permission is requested.
    /// It will be resolved with a boolean once the permission request completed.
    /// Only one pending permission result can be active at once.
    private var _pendingPermissionResult: FlutterResult?
    
    /// The pending Bluetooth on/off state result.
    ///
    /// This result is cached when the Bluetooth on/off state is queried on-demand
    /// and the state is either ``CBManagerState.unknown`` or ``CBManagerState.resetting``.
    /// It will be resolved with a boolean once ``CBManagerState.poweredOn`` or ``CBManagerState.poweredOff``
    /// is the current state.
    /// Only one pending Bluetooth state result result can be active at once.
    private var _pendingBluetoothIsOnOrOffResult: FlutterResult?
    
    public let bluetoothStateStreamHandler = BluetoothStateStreamHandler()
    
    public let bluetoothDiscoveryStreamHandler = BluetoothDeviceDiscoveryStreamHandler()
    
    /// Check whether Bluetooth permission is granted or denied.
    private func checkBluetoothPermissionStatus(bluetoothManager: CBCentralManager) -> Bool? {
        if #available(iOS 13.1, *) {
            switch(CBCentralManager.authorization) {
            case .notDetermined:
                return nil
            case .restricted:
                return false
            case .denied:
                return false
            case .allowedAlways:
                return true
            @unknown default:
                return false
            }
        } else if #available(iOS 13.0, *) {
            switch(bluetoothManager.authorization) {
            case .notDetermined:
                return nil
            case .restricted:
                return false
            case .denied:
                return false
            case .allowedAlways:
                return true
            @unknown default:
                return false
            }
        } else {
            // Below iOS 13.0, Bluetooh permissions do not exist.
            return true
        }
    }
    
    /// Handle changes in the state of the ``CBCentralManager``.
    func centralManagerDidUpdateState(_ central: CBCentralManager) {
        // Check if the permission result was pending and resolve it now.
        if(self._pendingPermissionResult != nil) {
            let permissionStatus = checkBluetoothPermissionStatus(bluetoothManager: central)
            
            // If the permission state is still undetermined, return false.
            self._pendingPermissionResult?(permissionStatus ?? false)
            self._pendingPermissionResult = nil
        }
        
        let bluetoothAdapterState = central.state

        self.bluetoothStateStreamHandler.onBluetoothStateChanged(state: bluetoothAdapterState)
        
        // Check if the Bluetooth on/off result was pending and resolve it now.
        if(self._pendingBluetoothIsOnOrOffResult != nil) {
            switch(bluetoothAdapterState) {
            case .unknown:
                break
            case .resetting:
                break
            case .unsupported:
                self._pendingBluetoothIsOnOrOffResult?(FlutterError(
                    code: BLUETOOTH_UNAVAILABLE_ERROR_CODE,
                    message: BLUETOOTH_UNAVAILABLE_ERROR_MESSAGE,
                    details: nil))
                self._pendingBluetoothIsOnOrOffResult = nil
            case .unauthorized:
                self._pendingBluetoothIsOnOrOffResult?(FlutterError(
                    code: BLUETOOTH_UNAUTHORIZED_ERROR_CODE,
                    message: BLUETOOTH_UNAUTHORIZED_ERROR_MESSAGE,
                    details: nil))
                self._pendingBluetoothIsOnOrOffResult = nil
            case .poweredOff:
                self._pendingBluetoothIsOnOrOffResult?(false)
                self._pendingBluetoothIsOnOrOffResult = nil
            case .poweredOn:
                self._pendingBluetoothIsOnOrOffResult?(true)
                self._pendingBluetoothIsOnOrOffResult = nil
            @unknown default:
                break
            }
        }
    }
    
    /// Handle Bleutooth device discovery results.
    func centralManager(_ central: CBCentralManager, didDiscover peripheral: CBPeripheral, advertisementData: [String : Any], rssi RSSI: NSNumber) {
        bluetoothDiscoveryStreamHandler.onDeviceFound(device: peripheral, advertisementData: advertisementData)
    }
    
    /// Get the current state of the Bluetooth adapter.
    func getBluetoothAdapterState(result: @escaping FlutterResult) {
        let state = _bluetoothManager?.state ?? CBManagerState.unsupported
        
        result(state.toSerializedString())
    }
    
    /// Initialize the ``CBCentralManager`` without showing a permission dialog.
    private func initializeBluetoothManager() {
        if(_bluetoothManager == nil) {
            // Allocating the `CBCentralManager` requests Bluetooth permissions.
            _bluetoothManager = CBCentralManager(delegate: self, queue: nil, options: [CBCentralManagerOptionShowPowerAlertKey: false])
            self.bluetoothStateStreamHandler.setCachedState(state: _bluetoothManager?.state)
        }
    }
    
    /// Check whether Bluetooth is currently on.
    func isBluetoothOn(result: @escaping FlutterResult) {
        guard let bluetoothManager = _bluetoothManager else {
            result(FlutterError(
                code: BLUETOOTH_UNAVAILABLE_ERROR_CODE,
                message: BLUETOOTH_UNAVAILABLE_ERROR_MESSAGE,
                details: nil))
            return
        }
        
        // If the current state is unknown or resetting,
        // cache the pending result and wait for a state change.
        switch(bluetoothManager.state) {
        case .unknown:
            _pendingBluetoothIsOnOrOffResult = result
        case .resetting:
            _pendingBluetoothIsOnOrOffResult = result
        case .unsupported:
            result(FlutterError(
                code: BLUETOOTH_UNAVAILABLE_ERROR_CODE,
                message: BLUETOOTH_UNAVAILABLE_ERROR_MESSAGE,
                details: nil))
        case .unauthorized:
            result(FlutterError(
                code: BLUETOOTH_UNAUTHORIZED_ERROR_CODE,
                message: BLUETOOTH_UNAUTHORIZED_ERROR_MESSAGE,
                details: nil))
        case .poweredOff:
            result(false)
        case .poweredOn:
            result(true)
        @unknown default:
            break
        }
    }
    
    /// Set up the ``CBCentralManager`` and request Bluetooth permissions.
    func requestBluetoothPermission(result: @escaping FlutterResult) {
        if(_pendingPermissionResult != nil) {
            result(FlutterError(
                code: BLUETOOTH_PERMISSION_REQUEST_ONGOING_ERROR_CODE,
                message: BLUETOOTH_PERMISSION_REQUEST_ONGOING_ERROR_MESSAGE,
                details: nil))
            
            return
        }
        
        _pendingPermissionResult = result
        
        // Set up the CBCentralManager if needed.
        initializeBluetoothManager()
        
        guard let bluetoothManager = _bluetoothManager else {
            result(FlutterError(
                code: BLUETOOTH_UNAVAILABLE_ERROR_CODE,
                message: BLUETOOTH_UNAVAILABLE_ERROR_MESSAGE,
                details: nil))
            return
        }
        
        let hasBluetoothPermision = checkBluetoothPermissionStatus(bluetoothManager: bluetoothManager)
        
        // Reset the pending result if the permission was granted or denied before.
        // Finally, resolve the permission if it is known.
        // If the permission is not known, it is resolved in `centralManagerDidUpdateState`.
        if(hasBluetoothPermision != nil) {
            result(hasBluetoothPermision)
            _pendingPermissionResult = nil
        }
    }
    
    /// Start a new Bluetooth scan.
    ///
    /// This method assumes that permission was granted before,
    /// since checking the authorization status might show a permission dialog.
    func startBluetoothScan(result: @escaping FlutterResult) {
        guard let bluetoothManager = _bluetoothManager else {
            result(FlutterError(
                code: BLUETOOTH_UNAVAILABLE_ERROR_CODE,
                message: BLUETOOTH_UNAVAILABLE_ERROR_MESSAGE,
                details: nil))
            return
        }
        
        if(!bluetoothManager.isScanning) {
            bluetoothManager.scanForPeripherals(withServices: [], options: [:])
        }
        
        result(nil)
    }
    
    /// Stop a running Bluetooth scan.
    ///
    /// This method assumes that permission was granted before,
    /// since checking the authorization status might show a permission dialog.
    func stopBluetoothScan(result: @escaping FlutterResult) {
        guard let bluetoothManager = _bluetoothManager else {
            result(FlutterError(
                code: BLUETOOTH_UNAVAILABLE_ERROR_CODE,
                message: BLUETOOTH_UNAVAILABLE_ERROR_MESSAGE,
                details: nil))
            return
        }
        
        if(bluetoothManager.isScanning) {
            bluetoothManager.stopScan()
        }

        result(nil)
    }
}

extension CBManagerState {
    func toSerializedString() -> String {
        switch(self) {
        case .unknown:
            return "unknown"
        case .resetting:
            return "resetting"
        case .unsupported:
            return "unavailable"
        case .unauthorized:
            return "unauthorized"
        case .poweredOff:
            return "off"
        case .poweredOn:
            return "on"
        @unknown default:
            return "unknown"
        }
    }
}

/// This class manages the Bluetooth device discovery ``FlutterEventSink``.
class BluetoothDeviceDiscoveryStreamHandler : NSObject, FlutterStreamHandler {
    private var sink: FlutterEventSink?
    
    func onDeviceFound(device: CBPeripheral, advertisementData: [String : Any]) {
        let deviceId: String = device.identifier.uuidString
        // If the device name is nil, check the advertisement packet.
        let deviceName: String? = device.name ?? advertisementData[CBAdvertisementDataLocalNameKey] as? String

        sink?([
            "deviceId": deviceId,
            "deviceName": deviceName,
        ])
    }
    
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        sink = events
        
        return nil
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        sink = nil
        
        return nil
    }
}

/// This class manages the stream of Bluetooth state changes.
class BluetoothStateStreamHandler : NSObject, FlutterStreamHandler {
    private var sink: FlutterEventSink?
    private var cachedBluetoothState: CBManagerState?
    
    func onBluetoothStateChanged(state: CBManagerState) {
        if(sink == nil) {
            cachedBluetoothState = state
        } else {
            sink?(state.toSerializedString())
        }
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        sink = nil
        
        return nil
    }
    
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        sink = events
        
        if(cachedBluetoothState != nil) {
            sink?(cachedBluetoothState?.toSerializedString())
        }
        
        return nil
    }
    
    func setCachedState(state: CBManagerState?) {
        cachedBluetoothState = state
    }
}

File: android/app/src/main/kotlin/be/weforza/app/BluetoothAdapterDelegate.kt
package be.weforza.app

import android.Manifest
import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothDevice
import android.bluetooth.le.ScanCallback
import android.bluetooth.le.ScanFilter
import android.bluetooth.le.ScanResult
import android.bluetooth.le.ScanSettings
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import androidx.core.app.ActivityCompat
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.EventChannel.StreamHandler
import io.flutter.plugin.common.MethodChannel

/**
 * This class represents the Bluetooth adapter implementation.
 */
class BluetoothAdapterDelegate(
    private val bluetoothAdapter: BluetoothAdapter?,
    val stateStreamHandler: BluetoothAdapterStateStreamHandler) {

    val deviceDiscoveryStreamHandler = BluetoothDeviceDiscoveryStreamHandler()

    /**
     * The Bluetooth scan callback for the running scan.
     * If this is not null, a scan is in progress.
     */
    private var bluetoothScanCallback: ScanCallback? = null

    companion object {
        const val BLUETOOTH_SCAN_RESULT_ERROR_CODE = "BLUETOOTH_SCAN_RESULT_ERROR"
        const val BLUETOOTH_SCAN_RESULT_ERROR_MESSAGE = "A scan result returned an error."

        const val BLUETOOTH_UNAUTHORIZED_ERROR_CODE = "BLUETOOTH_UNAUTHORIZED"
        const val BLUETOOTH_UNAUTHORIZED_ERROR_MESSAGE = "Access to Bluetooth was denied."

        const val BLUETOOTH_UNAVAILABLE_ERROR_CODE = "BLUETOOTH_UNAVAILABLE"
        const val BLUETOOTH_UNAVAILABLE_ERROR_MESSAGE = "Bluetooth is not supported on this device."
    }

    private fun buildBluetoothScanSettings(options: BluetoothScanOptions) : ScanSettings {
        val scanModes = mutableSetOf(
            ScanSettings.SCAN_MODE_BALANCED,
            ScanSettings.SCAN_MODE_LOW_LATENCY,
            ScanSettings.SCAN_MODE_LOW_POWER,
        )

        if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            scanModes.add(ScanSettings.SCAN_MODE_OPPORTUNISTIC)
        }

        val scanMode = if(scanModes.contains(options.scanMode)) {
            options.scanMode
        } else {
            ScanSettings.SCAN_MODE_BALANCED
        }

        var scanSettingsBuilder = ScanSettings.Builder().setScanMode(scanMode)

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            scanSettingsBuilder = scanSettingsBuilder.setPhy(ScanSettings.PHY_LE_ALL_SUPPORTED)
                .setLegacy(false)
        }

        return scanSettingsBuilder.build()
    }

    /**
     * Get the current state of the Bluetooth adapter.
     */
    fun getState(result: MethodChannel.Result) {
        if(bluetoothAdapter == null) {
            result.success("unavailable")
            return
        }

        try {
            when(bluetoothAdapter.state) {
                BluetoothAdapter.STATE_OFF -> result.success("off")
                BluetoothAdapter.STATE_ON -> result.success("on")
                BluetoothAdapter.STATE_TURNING_OFF -> result.success("turningOff")
                BluetoothAdapter.STATE_TURNING_ON -> result.success("turningOn")
                else -> result.success("unknown")
            }
        } catch (exception: SecurityException) {
            result.success("unauthorized")
        }
    }

    /**
     * Check if Bluetooth is currently on.
     *
     * Returns true if Bluetooth is currently on.
     * Returns false if Bluetooth is currently off.
     * Returns null otherwise.
     */
    fun isBluetoothOn(result: MethodChannel.Result) {
        if(bluetoothAdapter == null) {
            result.error(BLUETOOTH_UNAVAILABLE_ERROR_CODE, BLUETOOTH_UNAVAILABLE_ERROR_MESSAGE, null)
            return
        }

        try {
            when(bluetoothAdapter.state) {
                BluetoothAdapter.STATE_OFF -> result.success(false)
                BluetoothAdapter.STATE_ON -> result.success(true)
                // Set the pending result if the state is indeterminate.
                BluetoothAdapter.STATE_TURNING_OFF -> stateStreamHandler.setPendingBluetoothStateResult(result)
                BluetoothAdapter.STATE_TURNING_ON -> stateStreamHandler.setPendingBluetoothStateResult(result)
            }
        } catch (exception: SecurityException) {
            result.error(BLUETOOTH_UNAUTHORIZED_ERROR_CODE, BLUETOOTH_UNAUTHORIZED_ERROR_MESSAGE, null)
        }
    }

    /**
     * Start a new Bluetooth scan.
     */
    fun startBluetoothScan(options: BluetoothScanOptions, result: MethodChannel.Result, context: Context) {
        if(bluetoothAdapter == null) {
            result.error(BLUETOOTH_UNAVAILABLE_ERROR_CODE, BLUETOOTH_UNAVAILABLE_ERROR_MESSAGE, null)
            return
        }

        if(bluetoothScanCallback != null) {
            result.success(null)
            return
        }

        val hasScanPermission = if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            ActivityCompat.checkSelfPermission(
                context,
                Manifest.permission.BLUETOOTH_SCAN
            ) == PackageManager.PERMISSION_GRANTED
        } else {
            true
        }

        if(hasScanPermission) {
            bluetoothScanCallback = BluetoothScanCallback(deviceDiscoveryStreamHandler::onDeviceFound)

            val filters = listOf<ScanFilter>()

            bluetoothAdapter.bluetoothLeScanner.startScan(filters, buildBluetoothScanSettings(options), bluetoothScanCallback)
            result.success(null)
        } else {
            result.error(BLUETOOTH_UNAUTHORIZED_ERROR_CODE, BLUETOOTH_UNAUTHORIZED_ERROR_MESSAGE, null)
        }
    }

    /**
     * Stop a running Bluetooth device scan.
     */
    fun stopBluetoothScan(result: MethodChannel.Result, context: Context) {
        if(bluetoothAdapter == null) {
            result.error(BLUETOOTH_UNAVAILABLE_ERROR_CODE, BLUETOOTH_UNAVAILABLE_ERROR_MESSAGE, null)
            return
        }

        val bluetoothScanner = bluetoothAdapter.bluetoothLeScanner

        if(bluetoothScanCallback == null) {
            result.success(null)
            return
        }

        // Permission should be granted at this point, since the scan already started.
        // This is just a sanity check to silence the hint for bluetoothScanner.stopScan().
        val hasScanPermission = if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            ActivityCompat.checkSelfPermission(
                context,
                Manifest.permission.BLUETOOTH_SCAN
            ) == PackageManager.PERMISSION_GRANTED
        } else {
            true
        }

        if(hasScanPermission) {
            bluetoothScanner.stopScan(bluetoothScanCallback)
            bluetoothScanCallback = null // Clean up the old scan callback along with its settings.
            result.success(null)
        } else {
            result.error(BLUETOOTH_UNAUTHORIZED_ERROR_CODE, BLUETOOTH_UNAUTHORIZED_ERROR_MESSAGE, null)
        }
    }
}

/**
 * This class represents the Bluetooth adapter state change stream handler.
 */
class BluetoothAdapterStateStreamHandler : StreamHandler {
    /**
     * The pending Bluetooth on/off state result.
     *
     * This result is cached when the Bluetooth on/off state is queried on-demand
     * and the state is either [BluetoothAdapter.STATE_TURNING_OFF]
     * or [BluetoothAdapter.STATE_TURNING_ON].
     *
     * It will be resolved with a boolean once [BluetoothAdapter.STATE_ON]
     * or [BluetoothAdapter.STATE_OFF] is the current state.
     * Only one pending Bluetooth state result result can be active at once.
     */
    private var pendingBluetoothIsOnOrOffResult : MethodChannel.Result? = null

    /**
     * The event sink that collects the Bluetooth adapter's state changes.
     */
    private var sink: EventChannel.EventSink? = null

    fun onStateChanged(state: Int) {
        when(state) {
            // Resolve the pending state result if needed.
            BluetoothAdapter.STATE_OFF -> {
                sink?.success("off")
                if(pendingBluetoothIsOnOrOffResult != null) {
                    pendingBluetoothIsOnOrOffResult?.success(false)
                    pendingBluetoothIsOnOrOffResult = null
                }
            }
            BluetoothAdapter.STATE_ON -> {
                sink?.success("on")
                if(pendingBluetoothIsOnOrOffResult != null) {
                    pendingBluetoothIsOnOrOffResult?.success(true)
                    pendingBluetoothIsOnOrOffResult = null
                }
            }
            BluetoothAdapter.STATE_TURNING_OFF -> sink?.success("turningOff")
            BluetoothAdapter.STATE_TURNING_ON -> sink?.success("turningOn")
            else -> sink?.success("unknown")
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        sink = events
    }

    override fun onCancel(arguments: Any?) {
        sink = null
    }

    /**
     * Set the pending Bluetooth state result.
     */
    fun setPendingBluetoothStateResult(result: MethodChannel.Result) {
        if(pendingBluetoothIsOnOrOffResult == null) {
            pendingBluetoothIsOnOrOffResult = result
        }
    }
}

/**
 * This class represents the BroadcastReceiver for Bluetooth state changes.
 */
class BluetoothAdapterStateBroadcastReceiver(private val onStateChanged: (state: Int) -> Unit) : BroadcastReceiver() {
    override fun onReceive(context: Context?, intent: Intent?) {
        if(intent == null || !intent.action.equals(BluetoothAdapter.ACTION_STATE_CHANGED)) {
            return
        }

        val bluetoothAdapterState = intent.getIntExtra(BluetoothAdapter.EXTRA_STATE,
            BluetoothAdapter.ERROR)

        onStateChanged(bluetoothAdapterState)
    }
}

/**
 * This class represents the Bluetooth device discovery stream handler.
 */
class BluetoothDeviceDiscoveryStreamHandler : StreamHandler {
    private var sink: EventChannel.EventSink? = null

    fun onDeviceFound(bluetoothDevice: BluetoothDevice) {
        try {
            sink?.success(mapOf(
                "deviceName" to bluetoothDevice.name,
                "deviceId" to bluetoothDevice.address
            ))
        } catch (exception: SecurityException) {
            // If Bluetooth permission is not granted, skip the scan results.
            sink?.error(
                BluetoothAdapterDelegate.BLUETOOTH_SCAN_RESULT_ERROR_CODE,
                exception.message?: BluetoothAdapterDelegate.BLUETOOTH_SCAN_RESULT_ERROR_MESSAGE,
                null)
        }
    }

    override fun onCancel(arguments: Any?) {
        sink = null
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        sink = events
    }
}

/**
 * This class represents the Bluetooth scan callback for a Bluetooth scan.
 */
private class BluetoothScanCallback(
    private val onDeviceFound: (device: BluetoothDevice) -> Unit) : ScanCallback() {

    override fun onScanResult(callbackType: Int, result: ScanResult?) {
        super.onScanResult(callbackType, result)

        result?.device?.let {
            onDeviceFound(it)
        }
    }
}

/**
 * This class represents the options for a Bluetooth scan.
 */
data class BluetoothScanOptions(val scanMode: Int)